### PR TITLE
[SYSTEMML-1479] Integrated DeConvolution and Concat layer in Caffe2DML

### PR DESCRIPTION
**Pull request by niketanpansare**
@prithvirajsen @dusenberrymw can you please review this PR ?

@bertholdreinwald End-to-end integration of a network using them (for example: Unet) is still remaining.

### CODE REVIEW

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala [hunk 1]
```diff
@@ -212,15 +344,131 @@ class Elementwise(val param:LayerParameter, val id:Int, val net:CaffeNetwork) ex
...
+   *
+   * Inputs:
+   *  - n_i * c_i * h * w for each input blob i from 1 to K.
+   *
+   * Outputs:
+   *  - out: Outputs, of shape 
+   *    - if axis = 0: (n_1 + n_2 + ... + n_K) * c_1 * h * w, and all input c_i should be the same.
+   *    - if axis = 1: n_1 * (c_1 + c_2 + ... + c_K) * h * w, and all input n_i should be the same.
+   */
+  override def forward(dmlScript: StringBuilder, isPrediction: Boolean): Unit = {
```
Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: We should just add a new `nn/layers/concat.dml` layer, rather than embed it in this Scala code.  That way, we can test it for correctness using the same grad checks the rest of the `nn` library uses, and can use it in other contexts.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala [hunk 2]
```diff
@@ -249,53 +497,217 @@ class SoftmaxWithLoss(val param:LayerParameter, val id:Int, val net:CaffeNetwork
 }
 
 class ReLU(val param:LayerParameter, val id:Int, val net:CaffeNetwork) extends CaffeLayer {
+  // TODO: Leaky ReLU: negative_slope [default 0]: specifies whether to leak the negative part by multiplying it with the slope value rather than setting it to 0.
```
Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Yeah, we should add this as another layer.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala [hunk 3]
```diff
@@ -249,53 +497,217 @@ class SoftmaxWithLoss(val param:LayerParameter, val id:Int, val net:CaffeNetwork
...
+   *  - padw: Padding for left and right sides.
+   *      A typical value is 0.
+   *
+   * Outputs:
+   *  - out: Outputs, of shape (N, C*Hout*Wout).
+   *  - Hout: Output height.
+   *  - Wout: Output width.
+   */
   override def forward(dmlScript:StringBuilder, isPrediction:Boolean) = 
     invokeForward(dmlScript, List[String](out, "ignoreHout_"+id, "ignoreWout_"+id),
```
Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: By the way, with the new IPA scalar replacement + constant folding updates I've made, it's no longer necessary to compute `Hout` and `Wout` within Caffe2DML.  Once the constant folding PR is merged, they will be computed and replaced with literals during the initial compilation, thus no performance hit.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala [hunk 4]
```diff
@@ -304,54 +716,292 @@ class MaxPooling(val param:LayerParameter, val id:Int, val net:CaffeNetwork) ext
...
+    invokeBackward(dmlScript, outSuffix, List[String]("dOut" + id, dWeight, dBias), 
+        dout, Hout, Wout, X, weight, bias, numChannels, Hin, Win, kernel_h, kernel_w, stride_h, stride_w, pad_h, pad_w)
+  // n * c_o * h_o * w_o, where h_o = (h_i + 2 * pad_h - kernel_h) / stride_h + 1 and w_o likewise.
+  override def outputShape = ( numChannels, Hout, Wout )
+  // -------------------------------------------------
+  def numChannels = bottomLayerOutputShape._1
+  def Hin = bottomLayerOutputShape._2
+  def Win = bottomLayerOutputShape._3
+  // Hout = strideh * (Hin-1) - 2*padh + Hf + out_padh
+  def Hout:String =  try {
```
Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Once the constant folding PR is in, you can just use the `Hout` and `Wout` returned from `forward`.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala [hunk 5]
```diff
@@ -304,54 +716,292 @@ class MaxPooling(val param:LayerParameter, val id:Int, val net:CaffeNetwork) ext
...
+  // pad (or pad_h and pad_w) [default 0]: specifies the number of pixels to (implicitly) add to each side of the input
+  def pad_h =   if(convParam.hasPadH) convParam.getPadH.toString 
+                   else if(convParam.getPadCount > 0)  convParam.getPad(0).toString 
+                   else "0"
+  def pad_w =   if(convParam.hasPadW) convParam.getPadW.toString 
+                   else if(convParam.getPadCount > 0)  convParam.getPad(0).toString 
+                   else "0"
+}
+
+class DeConvolution(val param:LayerParameter, val id:Int, val net:CaffeNetwork) extends CaffeLayer with HasWeight with HasBias {
```
Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: What do you think about updating this to, say, `Convolution2DTranspose`?  I know that Caffe used "deconvolution", but it would be nice to not perpetuate that incorrect naming.  Also, Caffe2 now uses "ConvTranspose" instead of "Deconvolution" --> https://caffe2.ai/docs/operators-catalogue.html#convtranspose.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala [hunk 6]
```diff
@@ -99,8 +99,20 @@ trait CaffeLayer extends BaseDMLGenerator {
...
+  // Assumption: the first variable of resultVariables is always dX
   def invokeBackward(dmlScript:StringBuilder, outSuffix:String, resultVariables:List[String],  arguments:String*):Unit = {
-    invoke(dmlScript, sourceFileName + "::", resultVariables.map(_ + outSuffix), "backward", arguments.toList)
+    invoke(dmlScript, sourceFileName + "::", resultVariables.map(_ + outSuffix), "backward", arguments.toList, false)
+    val bottomLayerIDs = net.getBottomLayers(param.getName).map(l => net.getCaffeLayer(l).id)
+    dmlScript.append("; ")
+    bottomLayerIDs.map(bottomLayerID => dmlScript.append( dX(bottomLayerID) + outSuffix + " = " + resultVariables(0) + outSuffix + "; "))
+    dmlScript.append("\n")
+  }
+  def assignDoutToDX(dmlScript:StringBuilder, outSuffix:String):Unit = {
```
Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: What is this used for?

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala [hunk 7]
```diff
@@ -190,8 +321,9 @@ class Scale(val param:LayerParameter, val id:Int, val net:CaffeNetwork) extends
   if(!param.getScaleParam.getBiasTerm) throw new LanguageException("Add \"scale_param { bias_term: true }\" to the layer " + param.getName)
   override def sourceFileName = null
   override def init(dmlScript: StringBuilder): Unit = {}
+  // TODO: Generalize this !!
   def forward(dmlScript: StringBuilder, isPrediction: Boolean): Unit = assign(dmlScript, out, X)
-  override def backward(dmlScript: StringBuilder, outSuffix:String): Unit = assign(dmlScript, dX + outSuffix, dout)
+  override def backward(dmlScript: StringBuilder, outSuffix:String): Unit = assignDoutToDX(dmlScript, outSuffix)
```
Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Shouldn't we either just throw errors for Caffe2DML layers that aren't implemented yet?  Otherwise, I'm afraid we'll accidentally mask issues.  This layer can be mapped to the [`scale_shift2d`](https://github.com/apache/incubator-systemml/blob/master/scripts/nn/layers/scale_shift2d.dml) layer.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala [hunk 8]
```diff
@@ -144,16 +156,135 @@ class Data(val param:LayerParameter, val id:Int, val net:CaffeNetwork, val numCh
...
+   *  - C: Number of input channels (dimensionality of input depth).
+   *
+   * Outputs:
+   *  - gamma: Scale parameters, of shape (C, 1).
+   *  - beta: Shift parameters, of shape (C, 1).
+   *  - ema_mean: Exponential moving average of the mean, of
+   *      shape (C, 1).
+   *  - ema_var: Exponential moving average of the variance, of
+   *      shape (C, 1).
+   */
```
Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Thanks for adding the docs everywhere.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1460/

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 1 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: I thought of adding concat.dml, but it is difficult to generalize it because we donot support arbitrary number of arguments to a UDF.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 3 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Hout and Wout also needs to be computed for reading .caffemodel file at compile time.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 1 above)
Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: How about we add it for 2 arguments, and then Caffe2DML can just call it repeatedly.  Similar to what is being done now, except nest calls to `concat::forward`.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 4 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Btw, does Hout  and Wout cause issue with parfor based approach ?

Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: No, because with parfor distrib sgd, the layers are called (forward & backward) within the body of the parfor loop, and only the gradients are aggregated.  Regardless, Hout & Wout will be the same for all parallel instances of the parfor.  Did you have any other possible issues in mind?

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 1 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: The backward might be inefficient in that case :(

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 4 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Since the output shape logic is not too complex, I would really prefer to keep it until we are satisfied with the performance of distributed training with and without Caffe2DML.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 5 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: I am OK with either naming scheme as long as we are consistent.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 2 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Sure, and Caffe2DML can redirect to the appropriate layer.

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 7 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Yes, I intend to support that in different PR. Just don't want to make this PR too large :)

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 6 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: Updated the documentation :)

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1464/

File: src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala (hunk 7 above)
Comment (niketanpansare) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: To map this layer to [scale_shift2d.dml](https://github.com/apache/incubator-systemml/blob/master/scripts/nn/layers/scale_shift2d.dml), we will have to remove following code from batch_norm2d.dml:
https://github.com/apache/incubator-systemml/blob/master/scripts/nn/layers/batch_norm2d.dml#L118-L119

Comment (dusenberrymw) [inline src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala]: No we don't.  Just leave `gamma` and `beta` in their initialized forms (1s and 0s, respectively), and don't update them when you receive their gradients from `batch_norm2d::backward`.  That way, those lines will simply remain identity ops.
